backoffcache.go

```diff
@@ -0,0 +1,259 @@
+package discovery
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-peerstore/addr"
+)
+
+// BackoffDiscovery is an implementation of discovery that caches peer data and attenuates repeated queries
+type BackoffDiscovery struct {
+	disc         discovery.Discovery
+	stratFactory BackoffFactory
```

Comment:
@vyzo just a heads up that I renamed `strat -> stratFactory` in `BackoffDiscovery` to better differentiate it from the `strat` instances.